src/signature-help.jl

```diff
@@ -0,0 +1,354 @@
+"""
+function kwname(a::JL.SyntaxTree; sig=false)
+    if kind(a) === K"Identifier"
+        return a
+    elseif kind(a) === K"=" && kind(a[1]) === K"Identifier"
+        return a[1]
+    elseif sig && kind(a) === K"=" && kind(a[1]) === K"::" && kind(a[1][1]) === K"Identifier"
+        return a[1][1]
+    elseif kind(a) === K"..."
+        return nothing
+    end
+    JETLS_DEV_MODE && @info "Unknown kwarg form" a
+    return nothing
+end
+
+"""
+Best-effort mapping of kwname to position in `args`.  args[kw_i] and later are
+after the semicolon.  False negatives are fine here; false positives would hide
+signatures.
+
+If `sig`, then K"=" trees before the semicolon should be interpreted as optional
+positional args instead of kwargs.
+"""
+function find_kws(args::Vector{JL.SyntaxTree}, kw_i::Int; sig=false)
+    out = Dict{String, Int}()
+    for i in (sig ? (kw_i:lastindex(args)) : eachindex(args))
+        (kind(args[i]) != K"=") && i < kw_i && continue
+        n = kwname(args[i]; sig)
+        if !isnothing(n)
+            out[n.name_val] = i
+        end
+    end
+    return out
+end
+
+"""
+Information from one call's arguments for filtering signatures.
+- args: Every valid child of the K"call" and its K"parameters" if present
+- kw_i: One plus the number of args not in K"parameters" (semicolon)
+- pos_map: Map from position in `args` to (min, max) possible positional arg
+           e.g. f(a, k=1, b..., c) --> a => (1, 1), b => (2, nothing), c => (2, nothing)
+- kw_map: kwname => position in `args`
+
+TODO: types
+"""
+struct CallArgs
+    args::Vector{JL.SyntaxTree}
+    kw_i::Int
+    pos_map::Dict{Int, Tuple{Int, Union{Int, Nothing}}}
+    pos_args_lb::Int
+    pos_args_ub::Union{Int, Nothing}
+    kw_map::Dict{String, Int}
+end
+
+function CallArgs(st0::JL.SyntaxTree)
+    args, kw_i = flatten_args(st0)
+    pos_map = Dict{Int, Tuple{Int, Union{Int, Nothing}}}()
+    lb = 0; ub = 0
+    for i in eachindex(args[1:kw_i-1])
+        if kind(args[i]) === K"..."
+            ub = nothing
+            pos_map[i] = (lb, ub)
+        elseif kind(args[i]) != K"="
+            lb += 1
+            !isnothing(ub) && (ub += 1)
+            pos_map[i] = (lb, ub)
+        end
+    end
+    kw_map = find_kws(args, kw_i; sig=false)
+    CallArgs(args, kw_i, pos_map, lb, ub, kw_map)
+end
+
+"""
+Return false if we can definitely rule out `f(args...|` from being a call to `m`
+"""
+function compatible_call(m::Method, ca::CallArgs)
+    # TODO: (later) This should use type information from args (which we already
+    # have from m's params).  For now, just parse the method signature like we
+    # do in make_siginfo.
+
+    mstr = sprint(show, m)
+    mnode = JS.parsestmt(JL.SyntaxTree, mstr; ignore_errors=true)[1]
+
+    params, kwp_i = flatten_args(mnode)
+    has_var_params = kwp_i > 1 && kind(params[kwp_i - 1]) === K"..."
+    has_var_kwp = kwp_i <= length(params) && kind(params[end]) === K"..."
+
+    kwp_map = find_kws(params, kwp_i; sig=true)
+
```

Comment:
I feel that parsing the output of `sprint(show, m::Method)` and analyzing candidate methods based on it seems a bit hacky.
Is there a reason for taking this approach instead of using `Base.arg_decl_parts`/`Base.kwarg_decl` which are internally used by `Base.show_method`?
https://github.com/JuliaLang/julia/blob/9108dd08a572e394854d85aa0b2b680cc6a591c3/base/methodshow.jl#L217-L259

Reply:
As we discussed in the meeting, since we're using the output of `mstr = sprint(show, m::Method)` as `SignatureHelpInformation.label`, we need position information within `mstr` for parameter highlighting, which is why the current implementation is designed this way.

For future improvements, I think we could consider the following options:
- Adjust `Base.show_method` to add an option that removes the ` @ /path/to/method` part from the output (for making the parsing part safe)
- For `Method` objects that JETLS analyzes, maintain the original JS/JL definitions and use them for signature help as well

Anyway, I think we can proceed with this PR as it is for now.